lightning/src/routing/router.rs

```diff
@@ -150,39 +159,190 @@ impl cmp::PartialOrd for RouteGraphNode {
+		// For the last hop it will represent the value being transferred over this path.
+		for i in 0..self.hops.len() - 1 {
+			let mut next_hop_use_fee_msat = self.hops.get(i + 1).unwrap().hop_use_fee_msat;
+			let mut cur_hop = self.hops.get_mut(i).unwrap();
+			let transferred_amount_msat = cur_hop.next_hops_fee_msat + value_msat + next_hop_use_fee_msat;
+			// Overpay in fees if we can't save these funds due to htlc_minimum_msat.
+			// We try to account for htlc_minimum_msat in scoring (add_entry!), so that nodes don't
+			// set it too high just to maliciously take more fees by exploiting this
+			// match htlc_minimum_msat logic.
+			if let Some(extra_fees_msat) = cur_hop.htlc_minimum_msat.checked_sub(transferred_amount_msat) {
+				// If this is triggered, next_hops_fee_msat gets outdated.
+				next_hop_use_fee_msat += extra_fees_msat;
+			}
+			cur_hop.route_hop.fee_msat = next_hop_use_fee_msat;
+		}
+		self.hops.last_mut().unwrap().route_hop.fee_msat = value_msat;
+		self.hops.last_mut().unwrap().hop_use_fee_msat = 0;
+	}
+}
+
+fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
+	let proportional_fee_millions =
+		amount_msat.checked_mul(channel_fees.proportional_millionths as u64);
+	if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+			(channel_fees.base_msat as u64).checked_add(part / 1_000_000) }) {
+
+		Some(new_fee)
+	} else {
+		// This function may be (indirectly) called without any verification,
+		// with channel_fees provided by a caller. We should handle it gracefully.
+		None
+	}
+}
 
-/// Gets a route from us to the given target node.
+/// Gets a route from us (payer) to the given target node (payee).
 ///
 /// Extra routing hops between known nodes and the target will be used if they are included in
 /// last_hops.
 ///
 /// If some channels aren't announced, it may be useful to fill in a first_hops with the
 /// results from a local ChannelManager::list_usable_channels() call. If it is filled in, our
-/// view of our local channels (from net_graph_msg_handler) will be ignored, and only those in first_hops
-/// will be used.
+/// view of our local channels (from net_graph_msg_handler) will be ignored, and only those
+/// in first_hops will be used.
 ///
 /// Panics if first_hops contains channels without short_channel_ids
 /// (ChannelManager::list_usable_channels will never include such channels).
 ///
 /// The fees on channels from us to next-hops are ignored (as they are assumed to all be
-/// equal), however the enabled/disabled bit on such channels as well as the htlc_minimum_msat
-/// *is* checked as they may change based on the receiving node.
-pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, target: &PublicKey, first_hops: Option<&[&ChannelDetails]>,
+/// equal), however the enabled/disabled bit on such channels as well as the
+/// htlc_minimum/maximum_msat *are* checked as they may change based on the receiving node.
```

Comment:
What is the non-msat `htlc_minimum`?

Reply:
Oh, I thought it reads as "htlc_minimum_msat/htlc_maximum_msat", just a shortened version. Will fix.

Reply:
Ah, I just didnt read close enough (in general humans don't read inner letters in a word, we match on the ends).

Reply:
But you're probably right, no real need to save the space here.
I'll update this along with a fix for the issue above.